src/GSoft.Authentication.ClientCredentialsGrant.Tests/IntegrationTests.cs

```diff
@@ -137,4 +139,58 @@ public async Task Real_Client_Server_Communication()
             cts.Cancel();
         }
     }
+
+    private sealed class TestServerClient : DelegatingHandler
+    {
+        private readonly TestServer _testServer;
+        private HttpClient? _testServerClient;
+
+        public TestServerClient(TestServer testServer)
+        {
+            this._testServer = testServer;
+        }
+
+        protected override async Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
+        {
+            this._testServerClient ??= this._testServer.CreateClient();
+
+            // Request has to be cloned since it has already gone through the httpclient wrapping this handler even though the request hasn't been sent yet.
+            var cloneRequest = await this.CloneHttpRequest(request, cancellationToken);
+
+            return await this._testServerClient.SendAsync(cloneRequest, cancellationToken);
+        }
+
+        private async Task<HttpRequestMessage> CloneHttpRequest(HttpRequestMessage request, CancellationToken cancellationToken)
+        {
+            var cloneRequest = new HttpRequestMessage(request.Method, request.RequestUri);
+            cloneRequest.Version = request.Version;
+
+            foreach (var (key, value) in request.Headers)
+            {
+                cloneRequest.Headers.TryAddWithoutValidation(key, value);
+            }
+
+            foreach (var (key, value) in request.Options)
+            {
+                cloneRequest.Options.TryAdd(key, value);
+            }
+
+            if (request.Content != null)
+            {
+                var ms = new MemoryStream();
+                await request.Content.CopyToAsync(ms, cancellationToken).ConfigureAwait(false);
+                ms.Position = 0;
+
+                var streamContent = new StreamContent(ms);
+                foreach (KeyValuePair<string, IEnumerable<string>> header in request.Content.Headers)
+                {
+                    streamContent.Headers.Add(header.Key, header.Value);
+                }
+
+                cloneRequest.Content = streamContent;
```

Comment:
Because we are buffering the request content for testing purposes, we can make the process simpler:

```suggestion
                var contentBytes = await request.Content.ReadAsByteArrayAsync(cancellationToken);
                cloneRequest.Content = new ByteArrayContent(contentBytes);

                foreach (var (key, value) in request.Content.Headers)
                {
                    cloneRequest.Content.Headers.TryAddWithoutValidation(key, value);
                }
```